File: java/exercise3/CircularlyLinkedList.java
package exercise3;

public class CircularlyLinkedList<E> {

    private static class Node<E> {
        private E element;
        private Node<E> next;

        public Node(E e, Node<E> n) {
            element = e;
            next = n;
        }

        public E getElement() {
            return element;
        }

        public Node<E> getNext() {
            return next;
        }

        public void setNext(Node<E> n) {
            next = n;
        }
    }

    private Node<E> tail = null;
    private int size = 0;

    public CircularlyLinkedList() {
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public E first() {
        if (isEmpty())
            return null;
        return tail.getNext().getElement();
    }

    public E last() {
        if (isEmpty())
            return null;
        return tail.getElement();
    }

    public void rotate() {
        if (tail != null)
            tail = tail.getNext();
    }

    public void addFirst(E e) {
        if (size == 0) {
            tail = new Node<>(e, null);
            tail.setNext(tail);
        } else {
            Node<E> newest = new Node<>(e, tail.getNext());
            tail.setNext(newest);
        }
        size++;
    }

    public void addLast(E e) {
        addFirst(e);
        tail = tail.getNext();
    }

    public E removeFirst() {
        if (isEmpty())
            return null;
        Node<E> head = tail.getNext();
        if (head == tail)
            tail = null;
        else
            tail.setNext(head.getNext());
        size--;
        return head.getElement();
    }

    public String toString() {
        if (tail == null)
            return "()";
        StringBuilder sb = new StringBuilder("(");
        Node<E> walk = tail;
        do {
            walk = walk.getNext();
            sb.append(walk.getElement());
            if (walk != tail)
                sb.append(", ");
        } while (walk != tail);
        sb.append(")");
        return sb.toString();
    }

    public CircularlyLinkedList<E> clone() {
        CircularlyLinkedList<E> cloneList = new CircularlyLinkedList<>();
        if (!this.isEmpty()) {
            Node<E> current = this.tail.getNext();
            cloneList.addLast(current.getElement());
            current = current.getNext();
            while (current != this.tail.getNext()) {
                cloneList.addLast(current.getElement());
                current = current.getNext();
            }
        }
        return cloneList;
    }

    public boolean sameSequence(CircularlyLinkedList<E> other) {
        if (this.size() != other.size()) {
            return false;
        }

        if (this.isEmpty() && other.isEmpty()) {
            return true;
        }

        Node<E> start = this.tail.getNext();
        for (int i = 0; i < this.size(); i++) {
            Node<E> currentSelf = start;
            Node<E> currentOther = other.tail.getNext();
            boolean match = true;
            for (int j = 0; j < this.size(); j++) {
                if (!currentSelf.getElement().equals(currentOther.getElement())) {
                    match = false;
                    break;
                }
                currentSelf = currentSelf.getNext();
                currentOther = currentOther.getNext();
            }
            if (match) {
                return true;
            }
            start = start.getNext();
        }

        return false;
    }

    public static void main(String[] args) {
        CircularlyLinkedList<String> originalList = new CircularlyLinkedList<>();
        originalList.addLast("MSP");
        originalList.addLast("ATL");
        originalList.addLast("BOS");
        System.out.println(originalList);

        CircularlyLinkedList<String> clonedList = originalList.clone();
        System.out.println(clonedList);

        originalList.addLast("ABC");
        System.out.println(originalList);
        clonedList.addLast("XYZ");
        System.out.println(clonedList);

        CircularlyLinkedList<String> list1 = new CircularlyLinkedList<>();
        list1.addLast("MSP");
        list1.addLast("ATL");
        list1.addLast("BOS");

        CircularlyLinkedList<String> list2 = new CircularlyLinkedList<>();
        list2.addLast("ATL");
        list2.addLast("BOS");
        list2.addLast("MSP");

        CircularlyLinkedList<String> list3 = new CircularlyLinkedList<>();
        list3.addLast("BOS");
        list3.addLast("MSP");
        list3.addLast("ATL");

        CircularlyLinkedList<String> list4 = new CircularlyLinkedList<>();
        list4.addLast("MSP");
        list4.addLast("BOS");
        list4.addLast("ATL");

        System.out.println(list1);
        System.out.println(list2);
        System.out.println(list3);
        System.out.println(list4);

        System.out.println(list1.sameSequence(list2));
        System.out.println(list1.sameSequence(list3));
        System.out.println(list1.sameSequence(list4));
    }
}

File: src/exercise3/circularly_linked_list.py
# -*- coding: utf-8 -*-


class Node:
    def __init__(self, element, next_node=None):
        self.element = element
        self.next_node = next_node


class CircularlyLinkedList:
    def __init__(self):
        self.tail = None
        self.size = 0

    def __len__(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def first(self):
        if self.is_empty():
            return None
        return self.tail.next_node.element

    def last(self):
        if self.is_empty():
            return None
        return self.tail.element

    def rotate(self):
        if self.tail is not None:
            self.tail = self.tail.next_node

    def add_first(self, e):
        if self.is_empty():
            self.tail = Node(e)
            self.tail.next_node = self.tail  # a new list of one element
        else:
            newest = Node(e)
            newest.next_node = self.tail.next_node
            self.tail.next_node = newest
        self.size += 1

    def add_last(self, e):
        self.add_first(e)
        self.tail = self.tail.next_node

    def remove_first(self):
        if self.is_empty():
            return None
        old_head = self.tail.next_node
        if self.size == 1:
            self.tail = None
        else:
            self.tail.next_node = old_head.next_node
        self.size -= 1
        return old_head.element

    def __str__(self):
        if self.is_empty():
            return "[]"
        result = []
        current = self.tail.next_node
        result.append(str(current.element))
        current = current.next_node
        while current != self.tail.next_node:
            result.append(str(current.element))
            current = current.next_node
        return "[" + ", ".join(result) + "]"

    def clone(self):
        """
        Creates a clone of the circularly linked list.
        """
        # Initialize an empty clone list.
        clone_list = CircularlyLinkedList()

        # If the current list is not empty
        if not self.is_empty():

            # Add the element after tail to the clone list
            current = self.tail.next_node
            clone_list.add_last(current.element)

            # Traverse other elements and add them to the clone list
            current = current.next_node
            while current != self.tail.next_node:
                clone_list.add_last(current.element)
                current = current.next_node

        # Return the cloned list
        return clone_list

    def same_sequence(self, other):
        """
        Checks if two circularly linked lists have the same sequence of elements.
        """
        # Return false if the lengths of the two lists are different
        if len(self) != len(other):
            return False

        # Return true if both lists are empty
        if self.is_empty() and other.is_empty():
            return True

        # Start from the element after tail
        start = self.tail.next_node

        # Loop through each possible starting point in the current list
        for _ in range(len(self)):

            # Compare the elements of the two lists starting from the current position
            current_self = start
            current_other = other.tail.next_node
            match = True
            for _ in range(len(self)):
                if current_self.element != current_other.element:
                    match = False
                    break
                current_self = current_self.next_node
                current_other = current_other.next_node

            # If all elements match, the sequences are the same
            if match:
                return True

            # Move to the next starting point in the current list
            start = start.next_node

        # Return False if no matching sequence is found
        return False
